src/monitoring_service/database.py

```diff
@@ -1,51 +1,229 @@
+
+    def channel_count(self) -> int:
+        return self.conn.execute("SELECT count(*) FROM channel").fetchone()[0]
+
+
+class Database(SharedDatabase):
+    """ Holds all MS state which can't be quickly regenerated after a crash/shutdown """
+
+    def __init__(
+        self,
+        filename: str,
+        chain_id: int,
+        msc_address: str,
+        registry_address: str,
+        receiver: str,
+    ) -> None:
+        super(Database, self).__init__(filename)
+        self._setup(chain_id, msc_address, registry_address, receiver)
+
+    def _setup(
+        self,
+        chain_id: int,
+        msc_address: str,
+        registry_address: str,
+        receiver: str,
+    ) -> None:
+        """ Make sure that the db is initialized an matches the given settings """
+        assert chain_id >= 0
+        assert is_checksum_address(msc_address)
+        assert is_checksum_address(registry_address)
+        assert is_checksum_address(receiver)
+
+        initialized = self.conn.execute(
+            "SELECT name FROM sqlite_master WHERE type='table' AND name='blockchain'",
+        ).fetchone()
+        settings = [chain_id, msc_address, registry_address, receiver]
+
+        if initialized:
+            old_settings = self.conn.execute("""
+                SELECT chain_id,
+                       monitor_contract_address,
+                       token_network_registry_address,
+                       receiver
+                FROM blockchain
+            """).fetchone()
+            assert tuple(old_settings) == settings, f"{tuple(old_settings)} != {settings}"
+        else:
+            with self.conn:
+                # create db schema
+                with open(SCHEMA_FILENAME) as schema_file:
+                    self.conn.executescript(schema_file.read())
+                self.conn.execute("""
+                    UPDATE blockchain
+                    SET chain_id = ?,
+                        monitor_contract_address = ?,
+                        token_network_registry_address = ?,
+                        receiver = ?;
+                """, settings)
+
+    def update_state(self, state: MonitoringServiceState) -> None:
+        with self.conn:
+            self.conn.execute(
+                "UPDATE blockchain SET latest_known_block = ?",
+                [state.blockchain_state.latest_known_block],
+            )
+            self.conn.execute("DELETE FROM token_network")
+            self.conn.executemany(
+                "INSERT INTO token_network VALUES (?)",
+                [[address] for address in state.blockchain_state.token_network_addresses],
+            )
+
+    def load_state(self, sync_start_block: int) -> MonitoringServiceState:
+        """ Load MS state from db or return a new empty state if not saved one is present
 
-        return None
+        An empty state is initialized with `latest_known_block =
+        sync_start_block - 1`. If a saved state is present, `sync_start_block`
```

Comment:
Why the - 1 here? Won't that create problems during the first sync, where `sync_start_block` is 0?

Reply:
In `get_events`, we skip the skip `chain_state.latest_known_block` and continue with the next one, so we need `-1` to actually read all blocks after *and including* `sync_start_block`. Unfortunately, we have no tests yet that check that these boundaries are handled correctly.

Reply:
Makes sense, I guess then we need a `max(0, sync_start_block - 1)`

Reply:
If we put the max here, it won't work. Then latest_known_block will be 0 and we'll start processing with block 1. We can add it inside `get_events` or just skip it because the values passed in via command line can't be smaller than. If we put it inside `get_events`, I would prefer an error to silently using 0, instead.